lib/component.js

```diff
@@ -805,6 +826,89 @@ class Component extends WebComponent {
     return this._attrs;
   }
 
+  /**
+   * parse and validate config.params and create a param schema on the component
+   */
+  _syncParams() {
+    // maintain local validated map where all schema keys are defined
+    this._paramSchemas = {};
+    const paramSchemas = this.getConfig(`params`);
+    const defaultParams = this.getConfig(`defaultParams`);
+
+    for (const param of Object.keys(paramSchemas)) {
+      // convert type shorthand to object
+      let paramSchema = paramSchemas[param];
+      if (!paramSchema.type) {
+        paramSchema = {type: paramSchema};
+      }
+
+      // Ensure param type is valid
+      const type = paramSchema.type;
+      if (!PARAM_TYPE_DEFAULTS.has(type)) {
+        const typeString = typeof type === `function` ? type.name : String(type);
+        throw new Error(
+          `Invalid type: ${typeString} for param: ${param} in paramSchema. ` +
+            `Only (${Array.from(PARAM_TYPE_DEFAULTS.keys())
+              .map((v) => `'${v.name}'`)
+              .join(` | `)}) is valid.`,
+        );
+      }
+
+      if (defaultParams[param] && paramSchema.required) {
+        throw new Error(`param '${param}' in ${this.constructor.name} cannot have both required and default`);
+      }
+
+      const paramSchemaObj = {
+        type,
+        default: defaultParams[param],
+        required: paramSchema.required ? paramSchema.required : false,
+      };
+
+      // set default value for the params
+      this.params[param] = paramSchemaObj.default;
+
+      this._paramSchemas[param] = paramSchemaObj;
+    }
+
+    return this._paramSchemas;
+  }
+
+  get params() {
+    return this._params;
+  }
+
+  setParams(params) {
+    // if nothing change, then we return in advance to avoid unecessary update
+    if (Object.keys(params) === 0) {
+      return;
+    }
+    if (!this.shouldComponentUpdate(params, this.state)) {
+      return;
+    }
+    const updateOptions = {
+      cascade: false,
+    };
+    for (const param in params) {
+      if (!this._paramSchemas[param]) {
+        throw new Error(`extra param '${param}' on ${this.constructor.name} is not defined in schema`);
+      }
+      const newParam = params[param];
+      if (this._paramSchemas[param].required && (newParam === undefined || newParam === null)) {
+        throw new Error(`param '${param}' in ${this.constructor.name} is required, ${newParam} passed in`);
+      }
+    }
+    const newParams = Object.assign({}, this.params, params);
```

Comment:
Like I mentioned before, we're not updating the existing params, we're replacing params each time.
```suggestion
    const newParams = Object.assign({}, params);
```